### PR TITLE
Random seed setting

### PR DESCRIPTION
**Pull request by DickJC123**
## Description ##

I observed failures in "nosetests test_operator_gpu.py:test_random" in the check for repeatable sequences given identical seeds.  These failures are due to a lack of synchronization (i.e. race) between the setting of the seed by the gpu and a subsequent request for random numbers.  The test failures were only occasional and occurred more repeatably on the older, slower GPUs.

The first commit of this PR adds two new tests designed to accentuate the problem.  The test_random.py:test_parallel_random_seed_setting test targets the 'ResourceParallelRandom' that was recently added and fails consistently on multiple generations of GPUs.  The test_random.py:test_random_seed_setting test targets the legacy 'ResourceRandom' and does not exhibit any failures, suggesting the race issue is only present in the new parallel random generator.

The second commit of the PR adds the needed synchronization to ResourceParallelRandom and remedies the failure of the included test.  Since no failure was ever seen in ResourceRandom, I refrained from adding a similar synchronization where none is apparently needed.  I can add such a synchronization blindly if desired.

@piiswrong @szha @yzhliu might have some interest in this PR.

### Essentials ###
- [X] Passed code style checking (`make lint`)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [X] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: tests/python/unittest/test_random.py [hunk 1]
```diff
@@ -197,6 +197,75 @@ def test_random():
     check_with_device(mx.context.current_context(), 'float64')
 
 
+# Set seed variously based on `start_seed` and `num_init_seeds`, then set seed finally to `final_seed`
+def set_seed_variously(init_seed, num_init_seeds, final_seed):
+    end_seed = init_seed + num_init_seeds
+    for seed in range(init_seed, end_seed):
+        mx.random.seed(seed)
```
Comment (marcoabreu) [inline tests/python/unittest/test_random.py]: What's the point of setting the seed multiple times? Do you expect a different behaviour depending on whether you set it once or multiple times?

**Comment by marcoabreu** [comment]
You're probably not done with coding the tests yet, but there was one interesting observation I've made: If you check the test results at http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/incubator-mxnet/detail/PR-9409/1/pipeline, you'll notice that not all Python-tests are failing - meaning that the behaviour of the test itself is inconsistent in between environments.

File: tests/python/unittest/test_random.py (hunk 1 above)
Comment (DickJC123) [reply tests/python/unittest/test_random.py]: The problem being fixed by the PR is that the cpu pushes gpu seed-setting kernels to the engine, then can launch random-number-consuming kernels without synchronization (perhaps to a different stream).  Pushing multiple seed-setting kernels before the ultimate desired one is one way of aggravating this race condition.

**Comment by DickJC123** [comment]
The CI failures of my initial PR were due to the sensitive nature of the test_uniform_generator test, and my adding of new tests altered the random numbers seen by that test.  I have diagnosed and fixed the issue with test_uniform_generator: it was trying to fill buckets with probability 0.2 from a range of float16 values, which can only get so close given the limited precision.

**Comment by sxjscience** [comment]
If the test fails too often, we can also change the success rate to 0.15 https://github.com/DickJC123/mxnet/blob/master/python/mxnet/test_utils.py#L1777

**Pull request merged**